File: app/components/lookbook/elements/tab_panel_group/tab_panel_group.rb
module Lookbook
  class TabPanelGroup < Lookbook::BaseComponent
    with_slot :panel do |name = nil, **kwargs|
      kwargs = kwargs.deep_merge({
        x: {init: "registerPanel"}
      })

      Lookbook::TabPanel.new(name: name, **kwargs)
    end

    def initialize(name:, **kwargs)
      raise "Missing tab panel group name" unless name.present?
      set_tag_data_attr("tab-panel-group-id", name)
    end
  end
end

File: app/components/lookbook/elements/tab_panel_group/tab_panel_group.js
import { getComponent } from "@helpers/alpine";
import Component from "@js/component";

export default Component("tabPanelGroup", () => {
  return {
    groupId: null,
    panels: [],
    selectedPanel: null,

    init() {
      this.groupId = this.$el.getAttribute("data-tab-panel-group-id");
    },

    registerPanel() {
      const panel = getComponent(this.$el);
      this.panels.push(panel);
    },

    selectPanel(id) {
      this.selectedPanel = id;
    },

    root: {
      ["@tab:selected.window"](event) {
        const { group, tab } = event.detail;
        if (group === this.groupId) {
          this.selectPanel(tab);
        }
      },
    },
  };
});
